### PR TITLE
add `state_dict`/`load_state_dict` to base `Callback`

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Part of #11429 : introduce Stateful Callbacks
part 1. of the [full implementation PR reference](https://github.com/PyTorchLightning/pytorch-lightning/pull/11887#issuecomment-1045650933)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/base.py [hunk 1]
```diff
@@ -282,24 +282,66 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
+
+        Args:
+            trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
+            pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
+            checkpoint: entire loaded checkpoint dictionary
+
+        Note:
+            Between v1.6 and v1.8, both ``on_load_checkpoint_new(checkpoint)`` and
+            ``on_load_checkpoint(callback_state)`` will be supported.
+            In v1.8 only ``on_load_checkpoint(checkpoint)`` will be supported.
```
Comment (jjenniferdai) [inline pytorch_lightning/callbacks/base.py:333]: may be better to **not** offer this - instead, just a deprecation warning for overridden `on_load_checkpoint` that `callback_state` will become `checkpoint` in v1.8

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py [hunk 2]
```diff
@@ -282,24 +282,51 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
     def on_exception(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", exception: BaseException) -> None:
         """Called when any trainer execution is interrupted by an exception."""
 
+    def state_dict(self) -> Dict[str, Any]:
+        """Called when saving a checkpoint, implement to generate callback state_dict.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/base.py:286]:
```suggestion
        """Called when saving a checkpoint, implement to generate callback's ``state_dict``.
```

File: pytorch_lightning/callbacks/base.py [hunk 3]
```diff
@@ -282,24 +282,51 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Called when loading a checkpoint, implement to reload callback state given callback state_dict.
+
+        Args:
+            state_dict: the callback state returned by ``state_dict``.
+        """
+        pass
+
     def on_save_checkpoint(
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py:301]: returning callback state here will be deprecated - currently in the "`Note:`" annotation, should this have a `.. deprecated::` annotation too?

[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/callbacks/base.py:301]: if it's going to be removed eventually in future, then I think yes.

[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py:301]: `on_save_checkpoint` will still be available, but returning state in `on_save_checkpoint` will not. switched to `.. deprecated::` annotation though to raise more visibility

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py [hunk 4]
```diff
@@ -282,24 +282,52 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
-    ) -> dict:
-        """Called when saving a model checkpoint, use to persist state.
+    ) -> Optional[dict]:
+        r"""
+        .. deprecated:: v1.6
+            Returning state in this callback hook was deprecated in v1.6 and will be removed in v1.8.
+            Use ``state_dict`` instead to return state. In v1.8 ``on_save_checkpoint`` can only return None.
+
+        Called when saving a checkpoint to give you a chance to store anything else you might want to save.
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py:309]:
```suggestion
        r"""
        Called when saving a checkpoint to give you a chance to store anything else you might want to save.
```

File: pytorch_lightning/callbacks/base.py [hunk 5]
```diff
@@ -282,24 +282,52 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
 
         Args:
             trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
             pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
             checkpoint: the checkpoint dictionary that will be saved.
 
         Returns:
-            The callback state.
+            None or the callback state. Support for returning callback state will be removed in v1.8.
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py:317]:
```suggestion
            None or the callback state. Support for returning callback state will be removed in v1.8.
            
        .. deprecated:: v1.6
            Returning state in this callback hook was deprecated in v1.6 and will be removed in v1.8.
            Use ``state_dict`` instead to return state. In v1.8 ``on_save_checkpoint`` can only return None.
```

File: pytorch_lightning/callbacks/base.py [hunk 6]
```diff
@@ -282,24 +282,52 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
     def on_load_checkpoint(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", callback_state: Dict[str, Any]
     ) -> None:
-        """Called when loading a model checkpoint, use to reload state.
+        r"""
+        .. deprecated:: v1.6
+            This callback hook was deprecated in v1.6 and will be removed in v1.8. Use
+            ``load_state_dict`` instead. In v1.8 ``on_load_checkpoint(checkpoint)`` will replace
+            ``on_load_checkpoint(callback_state)``, i.e. input argument will be entire loaded
+            checkpoint dictionary instead of callback state.
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py:328]: just wondering why at the top? normally the first sentence is brief summary and this breaks the pattern

Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py:328]: was being consistent with the other deprecation annotations in `callbacks/base.py`?

File: pytorch_lightning/callbacks/base.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py:309]: agree with @Borda , the description of the hook should come first :)

File: pytorch_lightning/callbacks/base.py [hunk 7]
```diff
@@ -282,24 +282,52 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
 
     def on_load_checkpoint(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", callback_state: Dict[str, Any]
     ) -> None:
-        """Called when loading a model checkpoint, use to reload state.
+        r"""
+        .. deprecated:: v1.6
+            This callback hook was deprecated in v1.6 and will be removed in v1.8. Use
+            ``load_state_dict`` instead. In v1.8 ``on_load_checkpoint(checkpoint)`` will replace
+            ``on_load_checkpoint(callback_state)``, i.e. input argument will be entire loaded
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/base.py:327]: I'm confused by this message.
We won't remove the hook. What's deprecated is the signature only. If I understand correctly from the description, you will do the signature inspection + deprecation warning in a follow up PR. 

Notice: The deprecation (including all multi-part PRs) has to be completed before 1.6 or has to be dropped from 1.6 and added in 1.7.

Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py:327]: > We won't remove the hook. What's deprecated is the signature only. If I understand correctly from the description, you will do the signature inspection + deprecation warning in a follow up PR.

correct. do we want the deprecation annotations here? the different signatures will become BC so I added the annotation to raise more visibility

Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py:327]: Thanks for the confirmation. So I'm requesting to change:

> This callback hook was deprecated in v1.6 and will be removed in v1.8.

This cannot say that the hook gets removed. The hook will stay (the name is the same) but its signature and behavior will. 
IMO, a message like this would be more appropriate:

> This callback hook will change its signature and behavior in v1.8. 
> If you wish to load the state of the callback, use ``load_state_dict`` directly.
> In v1.8, ``on_load_checkpoint(checkpoint)`` will receive the entire loaded ....

Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py]: ahh got it, agree this message is better! updated

File: pytorch_lightning/callbacks/base.py (hunk 6 above)
Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py]: (going with at the bottom)

File: pytorch_lightning/callbacks/base.py [hunk 8]
```diff
@@ -282,24 +282,46 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
...
+        Args:
+            state_dict: the callback state returned by ``state_dict``.
+        """
+        pass
+
     def on_save_checkpoint(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
-    ) -> dict:
-        """Called when saving a model checkpoint, use to persist state.
+    ) -> Optional[dict]:
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/base.py:303]: nit: you could split this out to a separate PR so that this one is only about adding state_dict/load_state_dict

Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py:303]: just the signature change? move it to the PR that will check for returned state and warn?

Comment (ananthsub) [reply pytorch_lightning/callbacks/base.py:303]: yes, exactly. Depreating returning a value from `on_save_checkpoint` should be its own PR

File: pytorch_lightning/callbacks/base.py (hunk 7 above)
Comment (jjenniferdai) [reply pytorch_lightning/callbacks/base.py]: per below, leaving the `on_save/load_checkpoint` changes for a separate PR

**Review by ethanwharris — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
LGTM 
I assume the full changelog will come with the follow up?

**Comment by jjenniferdai** [comment]
> LGTM I assume the full changelog will come with the follow up?

yep, docs updates as well after all functionality is merged
